routing/dht/dht.go

```diff
@@ -279,9 +289,24 @@ func (dht *IpfsDHT) findProvidersSingle(ctx context.Context, p peer.ID, key key.
 
 // nearestPeersToQuery returns the routing tables closest peers.
 func (dht *IpfsDHT) nearestPeersToQuery(pmes *pb.Message, count int) []peer.ID {
-	key := key.Key(pmes.GetKey())
-	closer := dht.routingTable.NearestPeers(kb.ConvertKey(key), count)
-	return closer
+	if pmes.Key != nil {
+		return dht.routingTable.NearestPeers(kb.ConvertKey(key.Key(pmes.GetKey())), count)
+	} else {
+		closer := make(map[peer.ID]struct{})
+		for _, sk := range pmes.GetKeys() {
+			np := dht.routingTable.NearestPeers(kb.ConvertKey(key.Key(sk)), count)
+			for _, p := range np {
+				closer[p] = struct{}{}
+			}
+		}
+
+		var out []peer.ID
+		for p, _ := range closer {
+			out = append(out, p)
+		}
+
+		return out
```

Comment:
ahhhhh!!!! no, this is very weird, this is not "nearest peers to query" anymore, this returns a list of a bunch of nodes close to some random set of keys-- now a query could be uniformly distributed over the entire key-space, effectively saying "give me any peers in the table lol!". this is basically shooting a shotgun at the entire dht. 

i understand your point -- "this is already going to happen-- many of these peers will be contacted" and yes you're right, and you have a very good intuition to coalesce here, but [not like this](http://stream1.gifsoup.com/view1/4573874/not-like-this-o.gif), this wont scale the right way-- the right way is to either 
- (a) coalesce the keys that have very close XOR distances first before calling this function (i.e. cluster them), and then try this with clusters that have error below some known reasonable bound, or 
- (b) run this first for every key, and _then_ coalesce the messages going to the same target peers. (can do coalescing based on time, waiting even 250ms will likely coalesce a lot, and can keep doing it while the messages havent been sent yet, which is good given slow dials)

i think (b) is likely a lot easier